### PR TITLE
SourceDirectorySet => FileCollection

### LINKED ISSUE #219: SourceDirectorySet is too specific for combining multiple source files
Assigned to @confile for #216

### PR DESCRIPTION
**Pull request by brunobowden**
- Fixes #219
- Replace SourceDirectorySet with FileColleciton
- J2objcTestTaskTest unit tests for getTestNames()
- Minor comments

### CODE REVIEW

**Comment by brunobowden** [comment]
@advayDev1, @confile 

This one works, completing all the way through to the j2objcAssemble and then the j2objcTest task fails with a mysterious error:

```
Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: 2
	at com.github.j2objccontrib.j2objcgradle.tasks.J2objcTestTask$_testNames_closure2.doCall(J2objcTestTask.groovy:192)
	at com.github.j2objccontrib.j2objcgradle.tasks.J2objcTestTask.testNames(J2objcTestTask.groovy:180)
	at com.github.j2objccontrib.j2objcgradle.tasks.J2objcTestTask$testNames.callStatic(Unknown Source)
	at com.github.j2objccontrib.j2objcgradle.tasks.J2objcTestTask.test(J2objcTestTask.groovy:77)
	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:75)
	at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$StandardTaskAction.doExecute(AnnotationProcessingTaskFactory.java:226)
	at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$StandardTaskAction.execute(AnnotationProcessingTaskFactory.java:219)
	at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$StandardTaskAction.execute(AnnotationProcessingTaskFactory.java:208)
```

The IDE complains about some type mismatch. Figuring it out now...

**Comment by brunobowden** [comment]
I've figured out the problems. They're both related to PR #206.

Firstly the regular expression matcher was miscast to a String. Unfortunately the IDE doesn't warn on this:
```
def matcher = (testName =~ namespaceRegex)      // Original
String matcher = (testName =~ namespaceRegex)   // Broken
Matcher matcher = (testName =~ namespaceRegex)  // Fixed Again in this PR
```

Secondly the `testNames` method was given a return type of FileCollection when it's returning a set of strings. So the FileCollection had a set of nonsense paths to non-existent files. Curiously enough this still worked when each element of the FileCollection was converted back to a string. Note the `return testName` which is of type String.

Broken Version:
```
static FileCollection testNames(Project proj, FileCollection srcFiles, List<String> translateArgs) {
    Properties prefixesProperties = J2objcUtils.prefixProperties(proj, translateArgs)
    FileCollection testNames = srcFiles.collect { File file ->
        ...
        String testName = ...
        ...
        return testName
    }
    return testNames
}
```

Fixed Version:
```
static List<String> testNames(Project proj, FileCollection srcFiles, List<String> translateArgs) {
    Properties prefixesProperties = J2objcUtils.prefixProperties(proj, translateArgs)
    List<String> testNames = srcFiles.collect { File file ->
        ...
    }
    return testNames
}
```

**Comment by brunobowden** [comment]
@advayDev1, @confile - I've reworked this and added proper unit tests. PTAL

**Comment by confile** [comment]
LGTM

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTask.groovy [hunk 1]
```diff
@@ -50,7 +51,7 @@ class J2objcTestTask extends DefaultTask {
         return allFiles
     }
 
-    // Report of test failures
+    // Report of test failures, Gradle requires @Output for UP-TO-DATE checks
```
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTask.groovy]: i don't think this new comment increases clarity.
@OutputFile (not @Output) also does other things like letting you chain a CopyTask to the task to infer the source files for the CopyTask.  i'd get rid of the new comment.

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTask.groovy [hunk 2]
```diff
@@ -73,7 +74,8 @@ class J2objcTestTask extends DefaultTask {
 
         // list of test names: ['com.example.dir.ClassOneTest', 'com.example.dir.ClassTwoTest']
         // depends on "--prefixes dir/prefixes.properties" in translateArgs
-        FileCollection testNames = testNames(project, getSrcFiles(), getTranslateArgs())
+        Properties packagePrefixes = J2objcUtils.prefixProperties(project, translateArgs)
+        List<String> testNames = getTestNames(project, getTestSrcFiles(), packagePrefixes)
```
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTask.groovy]: nice refactor

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTask.groovy [hunk 3]
```diff
@@ -184,18 +185,19 @@ class J2objcTestTask extends DefaultTask {
...
+            Matcher matcher = (testName =~ /^(([^.]+\.)+)[^.]+$/)
             if (matcher.find()) {
                 String namespace = matcher[0][1]            // com.example.dir.
                 String namespaceChopped = namespace[0..-2]  // com.example.dir
-                if (prefixesProperties.containsKey(namespaceChopped)) {
-                    String value = prefixesProperties.getProperty(namespaceChopped)
-                    testName = testName.replace(namespace, value)
+                if (packagePrefixes.containsKey(namespaceChopped)) {
+                    String prefix = packagePrefixes.getProperty(namespaceChopped)
+                    testName = testName.replace(namespace, prefix)  // PrefixSomeTest
```
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTask.groovy]: this comment is a bit confusing.  is it talking about the value of testName at this statement or after this statement?

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcUtils.groovy [hunk 4]
```diff
@@ -52,15 +52,15 @@ class J2objcUtils {
...
 
         assert fileType == 'java' || fileType == 'resources'
         assert sourceSetName == 'main' || sourceSetName == 'test'
         SourceSet sourceSet = proj.sourceSets.findByName(sourceSetName)
-        // For standard fileTypes 'java' and 'resources,' per contract
-        // this cannot be null.
-        return sourceSet.getProperty(fileType)
+        // For standard fileTypes 'java' and 'resources,' per contract this cannot be null.
+        // Convert from SourceDirectorySet to FileCollection, see issue #216
+        return (FileCollection) sourceSet.getProperty(fileType)
```
Comment (advayDev1) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcUtils.groovy]: can you please try without this cast?  SourceDirectorySet inherits from FileCollection - it should not be needed.  all your other fixes should be sufficient.  (and also remove comment immediately above)

File: src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTaskTest.groovy [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+ */
+
+package com.github.j2objccontrib.j2objcgradle.tasks
+
+import org.gradle.api.InvalidUserDataException
+import org.gradle.api.Project
+import org.gradle.api.file.FileCollection
+import org.gradle.testfixtures.ProjectBuilder
+import org.junit.Before
+import org.junit.Test;
```
Comment (advayDev1) [inline src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTaskTest.groovy]: spurious semicolon

File: src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTaskTest.groovy [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    @Before
+    void setUp() {
+        proj = ProjectBuilder.builder().build()
+    }
+
+    @Test
+    public void testGetTestNames_Simple() {
+
+        // These are nonsense paths for files that don't exist
+        FileCollection srcFiles = proj.files ([
```
Comment (advayDev1) [inline src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTaskTest.groovy]: space between files and (

File: src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTaskTest.groovy [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                "${proj.rootDir}/src/test/java/com/example/parent/ParentClass.java",
+                "${proj.rootDir}/src/test/java/com/example/parent/subdir/SubdirClass.java",
+                "${proj.rootDir}/src/test/java/com/example/other/OtherClass.java"])
+        Properties noPackagePrefixes = new Properties()
+
+        List<String> testNames = J2objcTestTask.getTestNames(proj, srcFiles, noPackagePrefixes)
+
+        List<String> expectedTestNames = Arrays.asList(
```
Comment (advayDev1) [inline src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTaskTest.groovy]: just use standard groovy
List<String> eTN = ["a", "b", "c"]

File: src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTaskTest.groovy [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+    @Test
+    public void testGetTestNames_PackagePrefixes() {
+        Properties packagePrefixes = new Properties()
+        packagePrefixes.setProperty('com.example.parent', 'PrntPrefix')
+        packagePrefixes.setProperty('com.example.parent.subdir', 'SubPrefix')
+        packagePrefixes.setProperty('com.example.other', 'OthPrefix')
+
+        // These are nonsense paths for files that don't exist
+        FileCollection srcFiles = proj.files ([
```
Comment (advayDev1) [inline src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTaskTest.groovy]: ditto space

File: src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTaskTest.groovy [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+                "${proj.rootDir}/src/test/java/com/example/parent/ParentOneClass.java",
+                "${proj.rootDir}/src/test/java/com/example/parent/ParentTwoClass.java",
+                "${proj.rootDir}/src/test/java/com/example/parent/subdir/SubdirClass.java",
+                "${proj.rootDir}/src/test/java/com/example/other/OtherClass.java",
+                "${proj.rootDir}/src/test/java/com/example/noprefix/NoPrefixClass.java"])
+
+        List<String> testNames = J2objcTestTask.getTestNames(proj, srcFiles, packagePrefixes)
+
+        List<String> expectedTestNames = Arrays.asList(
```
Comment (advayDev1) [inline src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTaskTest.groovy]: ditto groovy idiom

**Comment by advayDev1** [comment]
commented.  LGTM with comments, if https://github.com/j2objc-contrib/j2objc-gradle/pull/220/files#r33089246 works out.  Otherwise let's keep discussing.

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTask.groovy (hunk 2 above)
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTask.groovy]: I'm much happier about this now

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTask.groovy (hunk 1 above)
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTask.groovy]: Remove comment

File: src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTask.groovy (hunk 3 above)
Comment (brunobowden) [inline src/main/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTask.groovy]: Added comment above clarifying that it's the LHS after assignment

File: src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTaskTest.groovy (hunk 5 above)
Comment (brunobowden) [inline src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTaskTest.groovy]: weird... removed

File: src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTaskTest.groovy (hunk 6 above)
Comment (brunobowden) [inline src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTaskTest.groovy]: removed space

File: src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTaskTest.groovy (hunk 7 above)
Comment (brunobowden) [inline src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTaskTest.groovy]: fixed

File: src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTaskTest.groovy (hunk 8 above)
Comment (brunobowden) [inline src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTaskTest.groovy]: fixed

File: src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTaskTest.groovy (hunk 9 above)
Comment (brunobowden) [inline src/test/groovy/com/github/j2objccontrib/j2objcgradle/tasks/J2objcTestTaskTest.groovy]: fixed

**Comment by brunobowden** [comment]
Removed cast and it all worked fine. Merging now.

**Comment by advayDev1** [comment]
@brunobowden  - Actually, you merged the wrong version of the commit.  Please revert, repush the correct commit, and remerge.

**Pull request merged**